extensions/snake/src/lib/game.ts

```diff
@@ -0,0 +1,269 @@
+        }
+        return result;
+    }
+}
+
+export class Game {
+    public field: Field = new Field();
+    private setField: React.Dispatch<React.SetStateAction<string>>;
+    private timeout: NodeJS.Timeout | undefined;
+    private snakeDirection: Move = Move.right;
+    public error: string | undefined;
+    public setError: React.Dispatch<React.SetStateAction<string | undefined>>;
+    public setMessage: React.Dispatch<React.SetStateAction<string | undefined>>;
+    public setScore: React.Dispatch<React.SetStateAction<GameScore | undefined>>;
+    private snake: Snake;
+    private foodCount = 0;
+    public speedMs = 1000;
+    public speed = 1;
+
+    constructor(setField: React.Dispatch<React.SetStateAction<string>>,
+        setError: React.Dispatch<React.SetStateAction<string | undefined>>,
+        setScore: React.Dispatch<React.SetStateAction<GameScore | undefined>>,
+        setMessage: React.Dispatch<React.SetStateAction<string | undefined>>) {
+        this.setField = setField;
+        this.setError = setError;
+        this.setScore = setScore;
+        this.setMessage = setMessage;
+        this.snake = new Snake({ x: 2, y: 2 });
+    }
+
+    public flush() {
+        this.setField(this.field.toString());
+    }
+
+    public start() {
+        this.field.clearField();
+        this.setError(undefined);
+        this.snake = new Snake({ x: getRandomInt(20, 80), y: getRandomInt(5, 15) });
+        this.foodCount = 0;
+        this.speed = 1;
+        this.speedMs = 200;
+        this.setScore({ food: this.foodCount, speed: 1 });
+        this.spawnFood();
+    }
+
+    public move(m: Move) {
+        let ok = false;
+        switch (m) {
+            case Move.right: {
+                if (this.snakeDirection !== Move.left) {
+                    ok = true;
+                }
+            } break;
+            case Move.left: {
+                if (this.snakeDirection !== Move.right) {
+                    ok = true;
+                }
+            } break;
+            case Move.up: {
+                if (this.snakeDirection !== Move.down) {
+                    ok = true;
+                }
+            } break;
+            case Move.down: {
+                if (this.snakeDirection !== Move.up) {
+                    ok = true;
+                }
+            } break;
+        }
+        if (ok) {
+            this.snakeDirection = m;
+        }
+    }
+
+    public increaseFood() {
+        this.foodCount += 1;
+        let newSpeedMs = this.speedMs - 20;
+        let addSpeed = true;
+        if (newSpeedMs < 10) {
+            newSpeedMs = 10;
+            addSpeed = false;
+        }
+        this.speedMs = newSpeedMs;
+        if (addSpeed) {
+            this.speed += 1;
+        }
+        this.setScore({ food: this.foodCount, speed: this.speed });
+        if (this.field.isFieldFull()) {
+            this.setMessage("You win 😀");
+        }
+    }
+
+    spawnFood() {
+        while (true) {
```

Comment:
This while loop will take a looong time when snake will be huge and board will be filled in 90% for example. A safer randomisation will be to generate list of potential positions and random one of them - single randomizatio that guarantees valid coordination. Otherwise we may end up looping here randomly forever until we hit valid position.

Reply:
Good spot. Add your suggestion. Thanks for the hint.